### PR TITLE
fix: source maps path on `windows`

### PR DESCRIPTION
**Pull request by evilebottnawi**
**What kind of change does this PR introduce?**
bugfix

**Did you add tests for your changes?**

yes

**If relevant, did you update the README?**

**Summary**

https://github.com/webpack-contrib/css-loader/issues/529

**Does this PR introduce a breaking change?**

No.

**Other information**

We need more tests and good review on this.

### CODE REVIEW

File: lib/loader.js [hunk 1]
```diff
@@ -22,6 +22,14 @@ module.exports = function(content, map) {
 	if(sourceMap) {
 		if (map && typeof map !== "string") {
 			map = JSON.stringify(map);
+
+			if (map.sourceRoot) {
+				map.sourceRoot = '';
+			}
+
+			if (map.sources) {
+				map.sources = map.sources.map((source) => source.replace(/\\/g, '/'));
```
Comment (michael-ciniawsky) [inline lib/loader.js]: `.map((source) => ...)` => `.map(function(source) { return ... })`

**Comment by michael-ciniawsky** [comment]
https://github.com/postcss/postcss/blob/master/lib/map-generator.es6#L188-L189

But it seems to work on the relative file path (L187) only atm and maybe unrelated 😛

File: lib/loader.js (hunk 1 above)
Comment (evilebottnawi) [inline lib/loader.js]: @michael-ciniawsky oh yep, i don't support old node.js in my project :smile:

Comment (michael-ciniawsky) [inline lib/loader.js]: Neither do I 😛 and I'm really favour make the switch as quickly as possible (node `< 4.0.0` is dead), but if we 'break' things all kind of conservatives will come out of their rat holes  and complain like crazy about it :D 'breaks app IN PRODUCTION' 'USE SEMVER!!!` etc...

File: lib/loader.js [hunk 2]
```diff
@@ -30,8 +39,8 @@ module.exports = function(content, map) {
 
 	processCss(content, map, {
 		mode: moduleMode ? "local" : "global",
-		from: loaderUtils.getRemainingRequest(this),
-		to: loaderUtils.getCurrentRequest(this),
+		from: loaderUtils.getRemainingRequest(this).split("!").pop(),
```
Comment (michael-ciniawsky) [inline lib/loader.js]: What is the output of `loaderUtils.getRemainingRequest(this).split("!").pop()`? If I remember right it should be the absolute file path and therefore we could replaced it with `this.resourcePath` instead?

File: lib/loader.js [hunk 3]
```diff
@@ -30,8 +39,8 @@ module.exports = function(content, map) {
...
+		to: loaderUtils.getCurrentRequest(this).split("!").pop(),
```
Comment (michael-ciniawsky) [inline lib/loader.js]: Should be equal to `this.resourcePath` aswell

File: lib/loader.js (hunk 2 above)
Comment (evilebottnawi) [inline lib/loader.js]: @michael-ciniawsky we can have `remainingRequest` https://github.com/webpack/loader-utils/blob/master/lib/getRemainingRequest.js#L4

Comment (evilebottnawi) [inline lib/loader.js]: Best get file from `getRemainingRequest` it is more safe

Comment (michael-ciniawsky) [inline lib/loader.js]: kk also not really related to this PR, just pop into my eyes

**Comment by evilebottnawi** [comment]
@michael-ciniawsky if we enable `appveyor` i can rewrite all tests for compatibility with `windows` (now tests written only for linux based os)

File: lib/loader.js [hunk 4]
```diff
@@ -20,8 +20,17 @@ module.exports = function(content, map) {
...
-		if (map && typeof map !== "string") {
-			map = JSON.stringify(map);
+		if (map) {
+			if (typeof map === "string") {
+				map = JSON.stringify(map);
+			}
+
+			if (map.sources) {
+				map.sources = map.sources.map(function (source) {
+					return source.replace(/\\/g, '/');
```
Comment (michael-ciniawsky) [inline lib/loader.js]: Test with `path.sep`, but I think it's unnecessary...

**Comment by michael-ciniawsky** [comment]
Yep appveyor would be good idea, want is need to enable it besides `appveyor.yml` ? For Org access ping bebraw or d3viant0ne 😛

File: lib/loader.js (hunk 4 above)
Comment (evilebottnawi) [inline lib/loader.js]: @michael-ciniawsky yep, we should already changed `\` to `/`

**Comment by evilebottnawi** [comment]
@michael-ciniawsky i think we should do this in `webpack-default` (no time for this :sob: )

**Comment by evilebottnawi** [comment]
@bebraw this https://github.com/webpack-contrib/css-loader/pull/532#discussion_r116264754 or https://github.com/webpack-contrib/css-loader/pull/532#discussion_r116258997 ?

File: lib/loader.js (hunk 4 above)
Comment (bebraw) [inline lib/loader.js]: @michael-ciniawsky This is a source map check. Can we assume forward slash there due to browser environment?

File: lib/loader.js (hunk 2 above)
Comment (bebraw) [inline lib/loader.js]: Look into `resourcePath`, yup. I can't remember the exact API, though. Maybe this needs some guidelines.

Comment (evilebottnawi) [inline lib/loader.js]: @bebraw why? what is bad use `loaderUtils.getRemainingRequest(this).split("!").pop()`, also https://github.com/webpack-contrib/css-loader/pull/532#discussion_r116260918

**Comment by d3viant0ne** [comment]
@bebraw - Need to check with Tobias direction before merging anything into css-loader, he specifically requested the code churn be kept to a minimum while he is working on new-loader.

**Comment by bebraw** [comment]
@d3viant0ne Yeah, fair point. Better get Tobias' approval first. 👍

File: lib/loader.js (hunk 2 above)
Comment (bebraw) [inline lib/loader.js]: @evilebottnawi I'm not entirely sure. At least test-wise the current implementation passes. It's fine to refactor later. 👍

**Comment by michael-ciniawsky** [comment]
@evilebottnawi @d3viant0ne @bebraw I think we can merge this meanwhile, despite the fact that current implementation will hopefully vanish soon 😛 , it's definitely a bug in the current `css-loader`

**Comment by d3viant0ne** [comment]
@evilebottnawi Open an issue in defaults for appveyor and we can get started on a org wide config as a part of defaults.

**Comment by d3viant0ne** [comment]
As far as this particular PR goes, there are a few open review comments. Patches are fine for the current major version so @bebraw @evilebottnawi @michael-ciniawsky if everyone is satisfied with this, throw a thumbs up on the comment and we'll get this landed.

**Comment by michael-ciniawsky** [comment]
AppVeyor Tracking Issue [#54](https://github.com/webpack-contrib/webpack-defaults/issues/54)

**Pull request merged**